sql,acceptance: Add scrub checks to schema changer and allocator tests

lgo opened the pull request:
This adds index integrity to all of the existing schema changer tests.
This compliments the existing assertions ensuring that the right amount
of KVs exist for the expected amount of rows.

This also adds additional checks using scrub to validate the integrity of
allocator acceptance tests when they run schema changes.

cc: @vivekmenezes were there any other tests you were thinking of that we could add these checks to?

lgo committed:
sql: Move scrub test helpers to testutils/sqlutils

lgo committed:
sql: Add scrub checks to schema changer tests

This adds index integrity to all of the existing schema changer tests.
This compliments the existing assertions ensuring that the right amount
of KVs exist for the expected amount of rows.

lgo committed:
acceptance: Add scrub checks to allocator tests

This also adds additional checks using scrub to validate the integrity
of allocator acceptance tests when they run schema changes.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20454)
<!-- Reviewable:end -->

lego commented:
All of the schema change tests currently have `checkTableKeyCount` calls to validate the size of the index.  @vivekmenezes thoughts on removing these completely and replacing them with a regular `COUNT(*)` row count check with a scrub check. That way we have the row count we expect and then scrub deals with the under the hood kv checks.

---

Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/acceptance/allocator_test.go, line 193 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20454#-L-Y6UXCDv9vBjhOiFzS:-L-Y6UXCDv9vBjhOiFzT:b-685ru5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1743620da3f2471bb71b9e37fc4422d6ba705903/pkg/acceptance/allocator_test.go#L193)):*
> ```Go
> 			// All these return the same result.
> 			validationQueries := []string{
> 				"SELECT COUNT(*) FROM tpch.customer AS OF SYSTEM TIME %s",
> ```

While these tests could have been removed, there are two things going on that I wasn't sure about:
1) There is a query on the introduced column, even though it's in the default family. Is this significant to check?
2) The queries are run at a specific timestamp. Is this required for this? (If so, there's an in-flight PR for adding AS OF SYSTEM TIME compatibility, so this will have to wait for that. #20368.)

---

*[pkg/sql/schema_changer_test.go, line 2434 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20454#-L-Y71edf8YsBjdX4bO5:-L-Y71eeqaw73ezMU5E3:b-vrw344) ([raw file](https://github.com/cockroachdb/cockroach/blob/1743620da3f2471bb71b9e37fc4422d6ba705903/pkg/sql/schema_changer_test.go#L2434)):*
> ```Go
> 	// FIXME(joey): It seems odd that we need to do this. As far as I
> 	// know, the knob is getting run because a new transaction is make
> 	// when running the following statement. Does this seem correct?
> ```

See the FIXME note. I wasn't quite certain if this is correct behaviour. As the note mentions, the scrub query seems to be run in it's own transaction. That causes the knob `SyncFilter` to be called again, which fails unless we nil out the channel. I believe this is an appropriate fix.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20454#-:-L-Y7PlaG3UBPIqS__N9:b-d3m43s)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on pkg/sql/schema_changer_test.go:
```diff
@@ -2372,6 +2425,17 @@ CREATE TABLE t.test (k INT PRIMARY KEY, v INT);
+	// FIXME(joey): It seems odd that we need to do this. As far as I
+	// know, the knob is getting run because a new transaction is make
+	// when running the following statement. Does this seem correct?
+	notifySchemaChange = nil
```
This is because the filter gets called even if the sql statement is not involved with a schema change. I'm not opposed to leave it the way you have.

vivekmenezes commented on pkg/testutils/sqlutils/scrub.go:
```diff
@@ -0,0 +1,87 @@
+}
+
+// GetScrubResultRows will scan and unmarshal ScrubResults from a Rows
+// iterator. The Rows iterate must from from an EXPERIMENTAL SCRUB
```
form from

vivekmenezes commented on pkg/testutils/sqlutils/scrub.go:
```diff
@@ -0,0 +1,87 @@
+	return results, nil
+}
+
+// RunScrub will run execute an exhaustive scrub check for a database.
```
for a table

vivekmenezes reviewed (approved):
Please run the acceptance tests before merging this PR. ```COCKROACH_EXTRA_FLAGS='-tf.keep-cluster=failed' build/teamcity-nightly-acceptance.sh TestRebalance_3To5Small_WithSchemaChanges```

lgo committed:
sql: Fix scrub physical checks when running against secondary indexes

Previously, an error would occur if the order of row data scanned in
MultiRowFetcher did not match the order of the index ColumnIDs. This
caused the wrong datums to get used during the physical check ordering
assertion.

Release note: None

vivekmenezes commented on pkg/sql/logictest/testdata/logic_test/scrub:
```diff
@@ -7,24 +7,19 @@ EXPERIMENTAL SCRUB TABLE t
 
-query TTTTTTTT
-EXPERIMENTAL SCRUB TABLE t WITH OPTIONS INDEX ALL
------
```
Why did you take this out?

The tests for this change are very subtle. Can we do something about that?

lego replied:
Welp, didn't mean to take that out. Added a short comment also mentioning what these are testing. These are testing more of the options than scrub, as the single run of scrub would be doing everything (so the rest is redundant).

vivekmenezes commented:
LGTM
